File: src/py/ipywidjucks/widgets/json.py
from .base import Base
from .base import T
from .base import W

import json


@W.register
class JSON(Base):
    _model_name = T.Unicode("JSONModel").tag(sync=True)

    source = T.Unicode("").tag(sync=True)
    value = T.Union([T.Dict(), T.List(), T.Unicode(), T.Int(), T.Float()]).tag(
        sync=True
    )

    @T.observe("source")
    def _source_changed(self, *_):
        self.value = json.loads(self.source)

File: packages/jupyterlab-wxyz/src/widgets/json.ts
import { FnModel } from './_base';

export class JSONModel extends FnModel<string, any, JSONModel.ITraits> {
  static model_name = 'JSONModel';

  defaults() {
    return {
      ...super.defaults(),
      _model_name: JSONModel.model_name,
      value: null as any
    };
  }

  async theFunction(source: string) {
    return JSON.parse(source);
  }
}

namespace JSONModel {
  export interface ITraits extends FnModel.ITraits<string, any> {}
}
